### ISSUE TITLE
Cleaner way to push custom backward pass down into vmap?

### ISSUE THREAD

**Issue by jeffgortmaker** [google/jax#2930]
This relates to my question in #2796 in that I'm using `custom_vjp` to replace some `nan` derivatives. Sorry about the long issue, this is sort of hard to explain. 

I'll slightly modify the `log1pexp` example in [this tutorial](https://jax.readthedocs.io/en/latest/notebooks/Custom_derivative_rules_for_Python_code.html#Example-problems) to illustrate my question. My objective function is `x \mapsto \sum_{y=1,50} \log(1 + \exp(xy))`.

```python
import jax
import jax.numpy as np
from functools import partial

f1 = lambda x, y: np.log(1. + np.exp(x * y))
g1 = jax.grad(lambda x: jax.vmap(partial(f1, x))(np.array([1., 50.])).sum())
print(g1(2.))
```

As in the tutorial, the gradient is `nan` at `x = 2` because the standard autodiff rule for `log1pexp` is numerically unstable when evaluated at `xy = 100`, which happens when `y = 50`. 

Unlike in the tutorial, I'm not going to define a more numerically stable `custom_jvp` for `log1pexp`. In my actual application this isn't as feasible. Instead, I'm going to try to replace the `nan` derivative in the sum with a  `0`. (My actual application is MLE with a large sum. In practice it's ok to zero out a few bad derivatives to keep the optimizer running.) The following function does this in the backward pass.

```python
clean = jax.custom_vjp(lambda x: x)
clean_fwd = lambda x: (x, None)
clean_bwd = lambda _, g: (np.where(np.isfinite(g), g, 0.),)
clean.defvjp(clean_fwd, clean_bwd)
```

Here's my first attempt at cleaning the derivative.

```python
f2 = lambda x, y: f1(clean(x), y)
g2 = jax.grad(lambda x: jax.vmap(partial(f2, x))(np.array([1., 50.])).sum())
print(g2(2.))
print(jax.make_jaxpr(g2)(2.))
```

I expected `clean` to zero out the `nan` derivative associated with `y = 50`, which would leave a gradient of `e^2 / (1 + e^2)`, approximately `0.88`. But the gradient here is `0`. Looking at the `jaxpr`, this is because internally, JAX applies my `clean` function *after* the sum, not within it.

I managed to get JAX to push `clean` into the `vmap` with a hackish `where` call that seems to "associate" `x` with each value of `y`.

```python
f3 = lambda x, y: f2(np.where(y, x, x), y)
g3 = jax.grad(lambda x: jax.vmap(partial(f3, x))(np.array([1., 50.])).sum())
print(g3(2.))
print(jax.make_jaxpr(g3)(2.))
```

This does what I wanted, but leaves me with two questions:

1. Any idea what's going on here? I don't really understand why my hack works, or how JAX decides whether to put an operation within or outside a `vmap`.
2. Is there a cleaner way of pushing `clean` into the `vmap`? Using `where` doesn't seem quite right.

**Comment by jekbradbury** [comment]
I think you have essentially the right diagnosis of what's going on. When you write
```python
f2 = lambda x, y: f1(clean(x), y)
f2v = lambda x: jax.vmap(partial(f2, x))(np.array([1., 50.])).sum()
```
we can expand this to
```python
def f2v(x):
  def inner(y):
    return f1(clean(x), y)
  return vmap(inner)(np.array([1., 50.])).sum()
```
Since `x` isn't an argument to the vmap, JAX runs `clean(x)` "outside the vmap" too and then broadcasts (adds a batch axis) to the result. But you want to clean the gradients inside the vmap, before summing them.

One way to do this is the hacky `np.where` you used. A more direct way to accomplish it would be to add a batch dimension to `x` and include it as a vmap argument:
```python
g3 = jax.grad(lambda x: jax.vmap(f2)(np.broadcast_to(x, (2,)), np.array([1., 50.])).sum())
```

I see it as an open question whether the behavior of `g4` using `in_axes=None` should be more like `g2` or more like `g3` (it currently behaves like `g2` with `vmap` and `g3` with `pmap`, which isn't great):
```python
g4 = jax.grad(lambda x: jax.vmap(f2, in_axes=(None, 0))(x, np.array([1., 50.])).sum())
```

Essentially, `vmap` implicitly introduces broadcasts; where those broadcasts go makes no difference on the forwards pass but can change the semantics of the backwards pass (since those broadcasts become sums). The behavior that surprised you is that the broadcasts related to a closed-over constant are introduced as late as possible (i.e., when that value is mixed with a batched value). This is probably not going to change, since it's a result of using data dependence/tracing to implement vmap. But, for `in_axes=None` values, should the broadcasts be introduced at the beginning of vmap or should they too happen as late as possible?

One last consideration is that this choice _doesn't matter_ unless you're using custom VJPs, because all backwards pass code synthesized by JAX is _linear_ (and therefore invariant to where we put broadcasts!). There's one other situation where the choice is user-visible: if the map is a `pmap`, then where the broadcasts go determines where and how many `psums` are introduced, and that could potentially have performance impacts (because the compiler tends not to optimize through them as well as it could with ordinary sums).

cc @mattjj since we talked about this recently; also, he likely has something to say about `custom_vjp` introducing nonlinear code into the backwards pass (this might be something we don't want to allow?).

**Comment by jeffgortmaker** [comment]
Thanks for taking a look and explaining! You cleared up all my questions.

I'm also curious about @mattjj's thoughts nonlinear `custom_vjp`'s. I think my use case is one example where this is actually super useful. I could also imagine that similar types of intermediate gradient clipping could be useful in lots of other settings. Perhaps there could be a better interface for introducing cleaning/clipping functionality.

**Comment by jekbradbury** [comment]
Matt clarified with me that we do want to enable nonlinear `rev` in `custom_vjp`s, and that in fact that's the reason we don't currently synthesize jvps from custom vjps by transposing `rev`. (If we were able to assume that `rev` is always linear, we'd be able to do that automatically, enabling forward-mode AD of functions with custom VJPs.)

But it looks like you've identified a case where we _are_ implicitly assuming that (or, more precisely, where we have two implementation options that give different results only if `rev` is nonlinear, but we didn't realize that before and didn't make a consistent choice).
